engine/table/src/main/java/io/deephaven/engine/table/impl/asofjoin/StaticAsOfJoinStateManagerTypedBase.java

```diff
@@ -0,0 +1,390 @@
+package io.deephaven.engine.table.impl.asofjoin;
+
+import io.deephaven.base.verify.Require;
+import io.deephaven.chunk.Chunk;
+import io.deephaven.chunk.ChunkType;
+import io.deephaven.chunk.attributes.Values;
+import io.deephaven.engine.rowset.*;
+import io.deephaven.engine.table.ColumnSource;
+import io.deephaven.engine.table.WritableColumnSource;
+import io.deephaven.engine.table.impl.sources.InMemoryColumnSource;
+import io.deephaven.engine.table.impl.sources.LongArraySource;
+import io.deephaven.engine.table.impl.sources.ObjectArraySource;
+import io.deephaven.engine.table.impl.sources.immutable.ImmutableObjectArraySource;
+import io.deephaven.engine.table.impl.util.TypedHasherUtil;
+import io.deephaven.engine.table.impl.util.TypedHasherUtil.BuildOrProbeContext.BuildContext;
+import io.deephaven.engine.table.impl.util.TypedHasherUtil.BuildOrProbeContext.ProbeContext;
+import org.apache.commons.lang3.mutable.MutableInt;
+import org.jetbrains.annotations.NotNull;
+
+import static io.deephaven.engine.table.impl.util.TypedHasherUtil.getKeyChunks;
+import static io.deephaven.engine.table.impl.util.TypedHasherUtil.getPrevKeyChunks;
+
+public abstract class StaticAsOfJoinStateManagerTypedBase extends StaticHashedAsOfJoinStateManager {
+    public static final int CHUNK_SIZE = 4096;
+    private static final long MAX_TABLE_SIZE = 1 << 30; // maximum array size
+    public static final Object EMPTY_RIGHT_STATE = null;
+
+    // the number of slots in our table
+    protected int tableSize;
+
+    protected long numEntries = 0;
+
+    // the table will be rehashed to a load factor of targetLoadFactor if our loadFactor exceeds maximumLoadFactor
+    // or if it falls below minimum load factor we will instead contract the table
+    private final double maximumLoadFactor;
+
+    // the keys for our hash entries
+    protected final ChunkType[] chunkTypes;
+    protected final WritableColumnSource[] mainKeySources;
+
+    protected final ImmutableObjectArraySource<RowSetBuilderSequential> leftRowSetSource; // use immutable
```

Comment:
comment unnecessary.  If we want a comment it would be describing which is the "primary" thing that determines whether the state is valid and what the sentinel values (if any are)